Fix possible division by zero in sys_spu_thread_write/read_ls, sys_spu_thread_un/bind_queue bugfix and stubs

elad335 opened the pull request:
* When type = 0, `lsa % type` was raising an exception. This is fixed and also move all EINVAL conditions before ESTAT and ESRCH checks as fw.
* Stub sys_spu_thread_group_log.
* Fix sys_spu_thread_un/bind_queue queue existence check. (check if binded, regardless of pointer experation)

Margen67 commented:
*division :^)

elad335 commented:
Fixed.

Nekotekina commented:
Using one function "valid()" for two pretty different tests isn't very good idea, because it's an unclear implication from pointer constness to desired test type. I doubt we need this helper in the first place.

elad335 commented:
Removed vm::ptr::valid commit.
Fixed sys_spu_thread_un/bind_queue, from hw tests it seems like it doesn't care if the underlying queue exists or not when checking for duplicates/finding it. instead it checks only if the queue number is registered or not.
[Testcase](https://github.com/elad335/myps3tests/tree/master/spu_tests/sys_spu_thread_bind_queue) code contains documentation as what the functions are testing.
sys_spu_thread_try/receive_event aren't affected.

master:
```
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error ok)
sys_event_queue_destroy(error ok)
sys_spu_thread_unbind_queue(error ESRCH)
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error ok)
sys_event_queue_destroy(error ok)
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error ok)
Reusing binded queue.
sys_event_queue_destroy(error ok)
sys_spu_thread_unbind_queue(error ESRCH)
SPU: sys_spu_thread_receive_event(error 0x80010002, data = {0xffffffff, 0xffffffff, 0xffffffff})
```

realhw/pr:
```
sys_event_queue_create(error ok)
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error ok)
sys_event_queue_destroy(error ok)
sys_spu_thread_unbind_queue(error ok)
sys_event_queue_create(error ok)
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error ok)
sys_event_queue_destroy(error ok)
sys_event_queue_create(error ok)
sys_event_queue_create(error ok)
sys_spu_thread_bind_queue(error EBUSY)
sys_spu_thread_bind_queue(error ok)
sys_event_queue_destroy(error ok)
sys_spu_thread_unbind_queue(error ok)
SPU: sys_spu_thread_receive_event(error 0x80010002, data = {0xffffffff, 0xffffffff, 0xffffffff})
```

Nekotekina commented on rpcs3/Emu/Cell/SPUThread.h:
```diff
@@ -570,7 +570,7 @@ class spu_thread : public cpu_thread
 	std::array<spu_int_ctrl_t, 3> int_ctrl; // SPU Class 0, 1, 2 Interrupt Management
 
-	std::array<std::pair<u32, std::weak_ptr<lv2_event_queue>>, 32> spuq; // Event Queue Keys for SPU Thread
+	std::array<std::pair<s64, std::weak_ptr<lv2_event_queue>>, 32> spuq; // Event Queue Keys for SPU Thread (-1 if not binded)
```
It's possible to test weak_ptr for nullptr: https://stackoverflow.com/questions/45507041/how-to-check-if-weak-ptr-is-empty-non-assigned

elad335 replied:
Thanks, changed it.

Nekotekina commented on rpcs3/Emu/Cell/lv2/sys_spu.cpp:
```diff
@@ -1219,29 +1219,33 @@ error_code sys_spu_thread_bind_queue(ppu_thread& ppu, u32 id, u32 spuq, u32 spuq
-	for (auto& v : thread->spuq)
-	{
-		if (v.second.expired())
+		if (target_num == spuq_num || v.second.lock() == queue)
```
Can also be compared without lock() by using owned_before

Nekotekina commented on rpcs3/Emu/Cell/lv2/sys_spu.cpp:
```diff
@@ -1219,29 +1219,34 @@ error_code sys_spu_thread_bind_queue(ppu_thread& ppu, u32 id, u32 spuq, u32 spuq
-	for (auto& v : thread->spuq)
-	{
-		if (v.second.expired())
+		if (v.first == spuq_num || v.second.owner_before(queue))
```
```suggestion
		if (v.first == spuq_num || (!v.second.owner_before(queue) && !queue.owner_before(v.second)))
```

elad335 replied:
Shouldn't the condition be inverted? it gives wrong results in testcase.

Nekotekina commented on rpcs3/Emu/Cell/lv2/sys_spu.cpp:
```diff
@@ -1219,29 +1219,34 @@ error_code sys_spu_thread_bind_queue(ppu_thread& ppu, u32 id, u32 spuq, u32 spuq
-				return CELL_EBUSY;
-			}
+			// Use the empty entry later
+			q = &v;
```
```suggestion
			if (!q)
				q = &v;
```
